### PR TITLE
Add Out of Sample Laplacian Spectral Embedding

### LINKED ISSUE #544: Add out-of-sample LSE
see #161

### PR DESCRIPTION
**Pull request by dtborders**
Fixes #544

Pull request to add out of sample lse to Graspologic. 

[Proof it works](https://nbviewer.jupyter.org/github/dtborders/archives/blob/master/Proof_It_Works.ipynb)

### CODE REVIEW

**Comment by kareef928** [comment]
Hey @dtborders, I can peer review this on Sunday, would that be fine?

File: graspologic/embed/lse.py
Comment (loftusa) [inline graspologic/embed/lse.py:243]: Since this code is so similar to the ASE oos code, I'm wondering if we can just refactor into the superclass to avoid code repitition?

File: docs/tutorials/embedding/OutOfSampleLSE.ipynb
Comment (loftusa) [inline docs/tutorials/embedding/OutOfSampleLSE.ipynb:1]: To be honest, it might be better to just have a single tutorial for out-of-sample embedding, called "Out-of-Sample Embedding", which just modifies the ASE one to add the LSE version. I think they're too similar to warrant separate tutorials

**Comment by loftusa** [comment]
(putting in my two cents as well, not to interfere with @kareef928's review)

**Comment by kareef928** [comment]
## Peer review checklist

### PR itself (on GitHub)
- [x] PR has a descriptive and succinct title.
- [x] PR has a _brief_ description of what was changed.
- [x] PR cites any related issues in the PR description (if applicable) and uses [closing keywords](https://docs.github.com/en/free-pro-team@latest/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) for any issues that should be closed as a result of the PR.
- [ ] PR does not have any extraneous file changes associated with it. One common example of an extraneous change is one that was already made in the target branch, but is showing up as part of this PR under `Files changed`. _is the test_sklearn file supposed to be changed?_
- [x] PR does not have any unresolved merge conflicts.
- [x] If implementing a major new feature or algorithm, a notebook demonstrating the proof of effectiveness is linked in the PR description. Depending on the feature, this kind of validation may take too long to run as a test or in a tutorial notebook. 

### Style
- [x] Code follows Python variable naming conventions, one description is [here](https://visualgit.readthedocs.io/en/latest/pages/naming_convention.html). Most variables should be `snake_case` and classes shoud be defined as `UpperCamelCase`, for example.
- [x] Code uses descriptive variable names throughout.
- [x] Code has no commented out "junk code."
- [x] Line lengths are short (recommended <88 chars). This includes docstrings. Note:  [black](https://black.readthedocs.io/en/stable/) tries to shorten line lengths but sometimes it is unable to do so automatically, especially for docstrings.

### Documentation
- [x] Any new public classes/functions have docstrings. 
- [x] Any new public classes/functions have been added to the appropriate `.rst` file [here](https://github.com/microsoft/graspologic/tree/dev/docs/reference) to be rendered by `Sphinx`.
- [x] Any major new features (like a new algorithm) are accompanied by a succinct tutorial notebook. Or, if it is agreed upon with the maintainers that this is out of scope for the current PR, a new issue has been created specifying that we need a tutorial notebook for this functionality.
- [x] Any new tutorial notebooks have been added to the appropriate folder [here](https://github.com/microsoft/graspologic/tree/dev/docs/tutorials) and included in the `.rst` file [here](https://github.com/microsoft/graspologic/blob/dev/docs/tutorial.rst) to be rendered by `Sphinx`.
- [x] Modifications to the documentation render appropriately in the `Netlify` build.

### Testing
- [x] New public classes/functions are tested to ensure they achieve the desired output.
- [x] New public classes/functions are tested to ensure proper errors are thrown when invalid inputs are passed.

File: docs/tutorials/embedding/OutOfSampleLSE.ipynb
Comment (dtborders) [reply docs/tutorials/embedding/OutOfSampleLSE.ipynb]: Just combined them!

File: graspologic/embed/lse.py
Comment (dtborders) [reply graspologic/embed/lse.py:173]: I moved transform to the BaseSpectralEmbed, but the fit methods are slightly different. Do we think it would be better to move the code that is the same in lse and ase fit to BaseSpectralEmbed, or would it be better to repeat a little bit of code?

File: docs/tutorials/embedding/OutOfSampleEmbed.ipynb
Comment (loftusa) [inline docs/tutorials/embedding/OutOfSampleEmbed.ipynb:25]: can you add a link to this tutorial

File: docs/tutorials/embedding/OutOfSampleEmbed.ipynb
Comment (loftusa) [inline docs/tutorials/embedding/OutOfSampleEmbed.ipynb:171]: "Similarly, we can also use Laplacian Spectral Embedding (LSE). We generate..."

File: docs/tutorials/embedding/OutOfSampleEmbed.ipynb
Comment (loftusa) [inline docs/tutorials/embedding/OutOfSampleEmbed.ipynb:353]: Weird question, but can you use a different random seed? The `LSE Out Latent Predictions` on my tutorial run has an out-of-sample latent position which is weirdly far away from the mean, so not very good as an explanatory image

File: docs/tutorials/embedding/OutOfSampleEmbed.ipynb
Comment (loftusa) [inline docs/tutorials/embedding/OutOfSampleEmbed.ipynb:428]: Whoa, your weighted plots look way, way off. You have a bug somewhere in the weighted case. Needs to be fixed before we can merge.

Screen Shot 2021-04-28 at 2 11 09 PM

File: graspologic/embed/lse.py
Comment (loftusa) [inline graspologic/embed/lse.py:6]: Could you add type-hinting for everything in this file? Trying to do that because the microsoft people want it in. See  
https://github.com/microsoft/graspologic/pull/772
https://github.com/microsoft/graspologic/pull/543

File: graspologic/embed/lse.py
Comment (loftusa) [reply graspologic/embed/lse.py:173]: imo, it's almost always better to avoid repeating code if possible, so i'd see if you can find a way to move the code that is the same up to the base class

File: tests/test_sklearn.py
Comment (loftusa) [inline tests/test_sklearn.py:9]: delete if not doing this

File: tests/test_spectral_embed.py
Comment (loftusa) [inline tests/test_spectral_embed.py:360]: out of curiosity, why are we dividing by `nodes_per_community`?

**Review by loftusa — COMMENTED** [review]
ok, overall looks fine. Biggest issue is that the weighted LSE in the tutorial is way, way off and we can't merge before that's fixed. Need type-hints, see the PRs I linked, and I'd see how much of `fit` you can refactor upwards. Otherwise nice job

**Comment by asaadeldin11** [comment]
@dtborders have you had a chance to address Alex's [comment](https://github.com/microsoft/graspologic/pull/722#discussion_r622424211)?

**Comment by dtborders** [comment]
@asaadeldin11 I have not figured it out yet, but I will be working on it tonight and hopefully have it by the morning.

File: docs/tutorials/embedding/OutOfSampleEmbed.ipynb
Comment (dtborders) [reply docs/tutorials/embedding/OutOfSampleEmbed.ipynb:25]: There is currently not a tutorial for LSE, do you want me to add the link to the ASE tutorial?

File: graspologic/embed/lse.py
Comment (dtborders) [reply graspologic/embed/lse.py:6]:
```
regularizer: Optional[float] = None
```
The regularizer can be an int or a float, is this correct?

File: docs/tutorials/embedding/OutOfSampleEmbed.ipynb
Comment (dtborders) [reply docs/tutorials/embedding/OutOfSampleEmbed.ipynb:428]: This has been fixed!

File: tests/test_spectral_embed.py
Comment (dtborders) [reply tests/test_spectral_embed.py]: That was a minor mistake and has been corrected.

File: graspologic/embed/lse.py
Comment (loftusa) [inline graspologic/embed/lse.py]: why were we originally dividing by (self.n_features_in_ / self.n_components_)?

File: graspologic/embed/lse.py
Comment (loftusa) [inline graspologic/embed/lse.py]: can do `X.T` instead of `np.transpose(X)`, probably cleaner

Comment (dtborders) [reply graspologic/embed/lse.py]: for sure

File: graspologic/embed/lse.py
Comment (dtborders) [reply graspologic/embed/lse.py]: Realized this was slightly off math wise, the current version is correct, they end up being about the same though

File: graspologic/embed/lse.py
Comment (loftusa) [reply graspologic/embed/lse.py:6]: can you link the place in the code where it is

Comment (dtborders) [reply graspologic/embed/lse.py:6]: https://github.com/microsoft/graspologic/blob/6ced01ec8ddd4e8c417fcd3a9f72b66f12ad39be/graspologic/embed/lse.py#L124

**Comment by dtborders** [comment]
@asaadeldin11 I have fixed all the problems with the code. I was wondering if you think I should try to avoid redundancy by somehow combining the similar parts of these two fit functions ([lse](https://github.com/microsoft/graspologic/blob/949dcb17e6b5d784d5ef5ee0fa92a8102aa23d62/graspologic/embed/lse.py#L138), [ase](https://github.com/microsoft/graspologic/blob/949dcb17e6b5d784d5ef5ee0fa92a8102aa23d62/graspologic/embed/ase.py#L139) , or should I leave it the way it currently is.

**Comment by asaadeldin11** [comment]
> @asaadeldin11 I have fixed all the problems with the code. I was wondering if you think I should try to avoid redundancy by somehow combining the similar parts of these two fit functions ([lse](https://github.com/microsoft/graspologic/blob/949dcb17e6b5d784d5ef5ee0fa92a8102aa23d62/graspologic/embed/lse.py#L138), [ase](https://github.com/microsoft/graspologic/blob/949dcb17e6b5d784d5ef5ee0fa92a8102aa23d62/graspologic/embed/ase.py#L139)) as discussed [here](https://github.com/microsoft/graspologic/pull/722#discussion_r622429329) or should I leave it the way it currently is.

yeah you can make a method in base that has this code if you have time.

File: graspologic/embed/lse.py
Comment (asaadeldin11) [reply graspologic/embed/lse.py]: @loftusa is there a reason why you put this at the end of fit rather than the beginning of transform? Seems like it's not a necessary calculation if someone doesn't have oos nodes.

Comment (loftusa) [reply graspologic/embed/lse.py]: mainly because it's an O(not much) set of calculations and I figured it might at some point be useful to have access to `self_pinv_left` prior to the transform, but I'm not at all opposed to changing it if you think it makes more sense

Comment (asaadeldin11) [reply graspologic/embed/lse.py]: that way we can put this all into transform instead of having the extra method in base. Also think it'll look cleaner.

Comment (dtborders) [reply graspologic/embed/lse.py]: Sounds good, ill move this all to transform

File: graspologic/embed/base.py
Comment (asaadeldin11) [inline graspologic/embed/base.py:291]:
```suggestion
    def _compute_oos_prediction(self, X, directed):
```

File: graspologic/embed/lse.py
Comment (asaadeldin11) [inline graspologic/embed/lse.py:168]:
```suggestion
    def _compute_oos_prediction(self, X, directed):
```

**Review by asaadeldin11 — CHANGES_REQUESTED** [review]
make these methods private so they don't show in the docs

**Comment by asaadeldin11** [comment]
@bdpedigo can we merge #716, since there seems to be changes overlapping with this

File: graspologic/embed/base.py
Comment (asaadeldin11) [inline graspologic/embed/base.py:5]:
```suggestion
import warnings

```

**Comment by dtborders** [comment]
@asaadeldin11 I believe this PR is all up to date with all the comments, is there anything else I can be working on for it?

File: graspologic/embed/base.py
Comment (asaadeldin11) [inline graspologic/embed/base.py:232]:
```suggestion
        out : np.ndarray OR length 2 tuple of np.ndarray
```

File: graspologic/embed/base.py
Comment (asaadeldin11) [inline graspologic/embed/base.py:235]:
```suggestion
            Array of latent positions, shape (n_oos_vertices, n_components) or (n_vertices, n_components). Transforms the fitted matrix if it was passed
            in.
```

File: graspologic/embed/lse.py
Comment (asaadeldin11) [inline graspologic/embed/lse.py:165]:
```suggestion
        """
        Computes the out-of-sample latent position estimation.
        Parameters
        ----------
        X: np.ndarray
            Input to do oos embedding on.
        directed: bool
            Indication if graph is directed or undirected
        Returns
        -------
        out : array_like or tuple, shape
```

File: graspologic/embed/ase.py
Comment (asaadeldin11) [inline graspologic/embed/ase.py:167]:
```suggestion
        """
        Computes the out-of-sample latent position estimation.
        Parameters
        ----------
        X: np.ndarray
            Input to do oos embedding on.
        directed: bool
            Indication if graph is directed or undirected
        Returns
        -------
        out : array_like or tuple, shape
        """
```

**Review by asaadeldin11 — COMMENTED** [review]
a few documentation comments then you should be good to go @dtborders

**Comment by dtborders** [comment]
@asaadeldin11  I just wanted to check in on this PR, is there anything else I can be working on for it?

**Review by asaadeldin11 — APPROVED** [review]

**Comment by dtborders** [comment]
@bdpedigo is there anything else I working on atm for this PR?

**Comment by bdpedigo** [comment]
not that I'm aware of, I just haven't had time to look over/merge

**Comment by dtborders** [comment]
sounds good, thanks!

File: graspologic/embed/ase.py
Comment (loftusa) [inline graspologic/embed/ase.py]: This commit made me realize it'd probably be smart to make a github action to just run black automatically every time a commit gets pushed. Not too familiar with github actions so I'm not sure how easy that'd be

**Comment by bdpedigo** [comment]
thanks @dtborders !

**Comment by dtborders** [comment]
Thanks a million!

**Pull request merged**
